components/indexer.py

```diff
@@ -72,37 +80,72 @@ def execute(self) -> None:
             self.intake_right_motor.set(0)
             return
 
-        if self.intaking:
-            injector = self.injector_motor
-            feeder = self.indexer_motors[-1]
-            intake_main_motor = self.intake_main_motor
+        injector = self.injector_motor
+        feeder = self.indexer_motors[-1]
+        intake_main_motor = self.intake_main_motor
+
+        if self.auto_retract:
+            if intake_main_motor.isFwdLimitSwitchClosed():
+                self.full_time += 1
+            else:
+                self.full_time = 0
+
+            if self.full_time >= self.full_timeout:
+                self.disable_intaking()
+                self.raise_intake()
 
-            if injector.isFwdLimitSwitchClosed():
-                self.transfer_to_injector = False
-            elif feeder.isFwdLimitSwitchClosed():
-                # Transferring
-                self.transfer_to_injector = True
+        if injector.isFwdLimitSwitchClosed():
+            self.transfer_to_injector = False
+        elif feeder.isFwdLimitSwitchClosed():
+            # Transferring
+            self.transfer_to_injector = True
 
-            # Turn on all motors and let the limit switches stop it
+        # find the location of the first ball
+        # only motors above it will run
+        if self.intaking:
+            first_ball = -1
+        else:
+            first_ball = len(self.all_motors)
+            if self.transfer_to_injector:
+                first_ball = len(self.all_motors) - 2
+                # there is a ball in the feeder
+            for i, motor in enumerate(self.all_motors):
+                if motor.isFwdLimitSwitchClosed():
+                    first_ball = i
+                    break
+
+        # Turn on all motors and let the limit switches stop it
+        if first_ball <= 0:
             intake_main_motor.set(self.intake_motor_speed)
-            for motor in self.indexer_motors:
+        else:
+            intake_main_motor.set(0)
+        for i, motor in enumerate(
+            self.indexer_motors, len(self.all_motors) - len(self.indexer_motors) - 1
```

Comment:
Would it be appropriate to just assume we will always have exactly one motor before the indexer motors? i.e. just pass `1` here. This is a little bit hard to read like this.